lib/libutee/tee_api_ree_service.c

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Intel Corporation
+ */
+
+#include <tee_api_types.h>
+#include <tee_api.h>
+#include <pta_generic.h>
+#include <string.h>
+
+
+struct __TEE_REESessionHandle {
+	uint64_t handle;
+	TEE_TASessionHandle session;
+};
+
+TEE_Result TEE_OpenREESession(REE_UUID *destination,
+			uint32_t cancellationRequestTimeout,
+			uint32_t paramTypes,
+			REE_Param params[REE_NUM_PARAMS],
+			TEE_REESessionHandle *ree_session,
+			uint32_t *returnOrigin)
+{
+	TEE_Param *pparams, iparam = {0};
+	TEE_UUID pta_generic = PTA_GENERIC_UUID;
+	TEE_Result result = TEE_SUCCESS;
+	TEE_REESessionHandle rsess;
+	TEE_Param init_params[REE_NUM_PARAMS];
+	uint32_t init_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+						TEE_PARAM_TYPE_VALUE_OUTPUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+
+	if (!destination || !ree_session || !returnOrigin)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rsess = TEE_Malloc(sizeof(*rsess), TEE_MALLOC_FILL_ZERO);
+	if (!rsess)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/* Open a session on the custom PTA */
+	result = TEE_OpenTASession(&pta_generic, 0, 0, NULL,
+					&rsess->session, NULL);
+	if (result != TEE_SUCCESS) {
+		MSG("Failed to open session on REE\n");
+		goto err;
+	}
+
+	/* Construct the init parameters */
+	memset(init_params, 0, sizeof(init_params));
+	init_params[0].memref.buffer = destination;
+	init_params[0].memref.size = sizeof(REE_UUID);
+	result = TEE_InvokeTACommand(rsess->session, 0, OPTEE_MRC_GENERIC_OPEN,
+				init_param_types, init_params, returnOrigin);
+	if (result != TEE_SUCCESS) {
+		MSG("Failed to find the ree service\n");
+		goto err;
+	}
+	rsess->handle = init_params[1].value.a;
```

Comment:
We could let the PTA save this information if we have each PTA session only hold one REE session. I don't expect many enough REE session to make that a problem.

Reply:
My thinking process was to attach the REE context to UTA, since, it initiated the call.

Reply:
Who cleans up when the TA panics or crashes?

Reply:
I have not seen the code where TA clean up is done. If handling session at PTA does the job, can you please guide me how to do that.

Reply:
It's something you can do in `pta_ree_service_close_session()`

Reply:
I need to explore that. If it can be taken in next patchset, please let me know. That will be really helpful.